### PR TITLE
Flag usage recommendations

### PR DESCRIPTION
**Pull request by thomasnyman**
In PR https://github.com/ossf/wg-best-practices-os-developers/pull/132#issuecomment-1505292993 @david-a-wheeler noted that we need to note somewhere that the option flags are order-sensitive. Another feedback from the C/C++ Compiler BP Guide calls was to add the recommendation to to explicitly enable hardening flag in a projects build system rather then relying on flags configured in the local GCC profile. This PR adds text to address these concerns.

### CODE REVIEW

**Comment by thomasnyman** [comment]
Updated PR with rebase with respect to main after renaming files

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md [hunk 1]
```diff
@@ -76,9 +76,15 @@ The recommendations in Table 1 and Table 2 are primarily applicable to compiling
...
 
-You can typically find the default flags used by GCC on your system by examining the output of the `gcc -v` command. Examining the `Configured with:` section, in the `gcc -v` output, can be useful for understanding how GCC is set up on your system and what options were chosen by the distribution's maintainers during the build process. It can also be useful for diagnosing issues with GCC or for troubleshooting problems that may arise during the compilation of software.
+You can typically find the default flags used by GCC on your system by examining the output of the `gcc -v` command. Examining the `Configured with:` section, in the `gcc -v` output, can be useful for understanding how GCC is set up on your system and what options were chosen by the distribution's maintainers during the build process. It can also be useful for diagnosing issues with GCC or for troubleshooting problems that may arise during the compilation of software. For example, issues caused by interactions between order-sensitive flags.
```
Comment (thomasnyman) [inline docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:79]: Should include the corresponding flag for Clang (and possible the Microsoft compiler?)

Possible also the relevant variant for debugging issues with macros (`gcc -O2 -dM -E - < /dev/null`)

**Comment by thomasnyman** [comment]
Minor further tweaks to avoid referring to only GCC.

**Review by gkunz — CHANGES_REQUESTED** [review]
generally looks good to me - except for the footnote syntax fix.

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md [hunk 2]
```diff
@@ -74,11 +74,19 @@ The recommendations in Table 1 and Table 2 are primarily applicable to compiling
...
+
+[^debian-hardening]: Software in the Public Interest, [Hardening in Debian](https://wiki.debian.org/Hardening), Debian Wiki
+[^gentoo-hardening]: Gentoo Foundation, [Hardening in Gentoo](https://wiki.gentoo.org/wiki/Hardened/Toolchain), Gentoo Wiki
+[^fedora-hardening]: Red Hat, [Using RPM build flags in Fedora](https://src.fedoraproject.org/rpms/redhat-rpm-config/blob/rawhide/f/buildflags.md), Fedora Package Sources (`redhat-rpm-config`)
+[^opensuse-hardening]: SUSE, [openSUSE Security Features](https://en.opensuse.org/openSUSE:Security_Features), openSUSE Wiki
+[^ubuntu-hardening]: Ubuntu, [Ubuntu Security Features](https://wiki.ubuntu.com/Security/Features), Ubuntu Wiki
```
Comment (gkunz) [inline docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:89]: Based on #209:

```suggestion
[^debian-hardening]: Software in the Public Interest, [Hardening in Debian](https://wiki.debian.org/Hardening), Debian Wiki

[^gentoo-hardening]: Gentoo Foundation, [Hardening in Gentoo](https://wiki.gentoo.org/wiki/Hardened/Toolchain), Gentoo Wiki

[^fedora-hardening]: Red Hat, [Using RPM build flags in Fedora](https://src.fedoraproject.org/rpms/redhat-rpm-config/blob/rawhide/f/buildflags.md), Fedora Package Sources (`redhat-rpm-config`)

[^opensuse-hardening]: SUSE, [openSUSE Security Features](https://en.opensuse.org/openSUSE:Security_Features), openSUSE Wiki

[^ubuntu-hardening]: Ubuntu, [Ubuntu Security Features](https://wiki.ubuntu.com/Security/Features), Ubuntu Wiki
```

**Comment by thomasnyman** [comment]
Squashed requested changes to original commit.

**Comment by thomasnyman** [comment]
Merged as per discussion in [C/C++ Compiler Guide BP call 2023-08-02](https://docs.google.com/document/d/1UClGUnOSkOH_wab6Lx43KUdkaK37L8sbWJ_GPZvc1YY/edit#bookmark=id.82homkohrlln).

**Pull request merged**